examples/ex9.cpp

```diff
@@ -51,6 +55,1094 @@ double inflow_function(const Vector &x);
+            }
+         }
+
+         ////////////////////////////
+         // Boundary contributions //
+         ////////////////////////////
+         if (dim==1)
+            numBdrs = 0; // Nothing needs to be done for 1D boundaries
+         else if (dim==2)
+            mesh->GetElementEdges(k, bdrs, orientation);
+         else if (dim==3)
+            mesh->GetElementFaces(k, bdrs, orientation);
+         
+         FillNeighborDofs(mesh, numDofs, k, nd, p, dim, bdrs);
+         
+         for (i = 0; i < numBdrs; i++)
+         {
+            Trans = mesh->GetFaceElementTransformations(bdrs[i]);
+            vn = 0.;
+            
+            for (l = 0; l < irF1->GetNPoints(); l++)
+            {
+               const IntegrationPoint &ip = irF1->IntPoint(l);
+               IntegrationPoint eip1;
+               Trans->Face->SetIntPoint(&ip);
+               
+               if (dim == 1)
+                  nor(0) = 2.*eip1.x - 1.0;
+               else
+                  CalcOrtho(Trans->Face->Jacobian(), nor);
+               
+               if (Trans->Elem1No != k)
+               {
+                  Trans->Loc2.Transform(ip, eip1);
+                  el.CalcShape(eip1, shape);
+                  Trans->Elem2->SetIntPoint(&eip1);
+                  coef.Eval(vval, *Trans->Elem2, eip1);
+                  nor *= -1.;
+                  Trans->Loc1.Transform(ip, eip1);
+                  el.CalcShape(eip1, shapeNeighbor);
+                  
+                  neighborElem = Trans->Elem1No;
+               }
+               else
+               {
+                  Trans->Loc1.Transform(ip, eip1);
+                  el.CalcShape(eip1, shape);
+                  Trans->Elem1->SetIntPoint(&eip1);
+                  coef.Eval(vval, *Trans->Elem1, eip1);
+                  Trans->Loc2.Transform(ip, eip1);
+                  el.CalcShape(eip1, shapeNeighbor);
+                  
+                  neighborElem = Trans->Elem2No;
+               }
+
+               nor /= nor.Norml2();
+               vn = min(0., vval * nor);
+               
+               for(j = 0; j < numDofs; j++)
+               {
+                  bdrIntLumped(k*nd+dofs(j,i),i) -= ip.weight * 
+                     Trans->Face->Weight() * shape(dofs(j,i)) * vn;
+                  
+                  for (m = 0; m < numDofs; m++)
+                  {
+                     bdrInt(k*nd+dofs(j,i),i*nd+dofs(m,i)) += ip.weight * 
+                        Trans->Face->Weight() * shape(dofs(j,i)) * shape(dofs(m,i)) * vn;
+                     
+                     if (neighborElem != 0)
+                        idx = ((int)(neighborDof(k*numDofs+m,i))) % (neighborElem*nd);
+                     else
+                        idx = neighborDof(k*numDofs+m,i);
+                     
+                     bdrIntNeighbor(k*nd+dofs(j,i),i*nd+dofs(m,i)) += ip.weight * 
+                        Trans->Face->Weight() * shape(dofs(j,i)) * shapeNeighbor(idx) * vn;
```

Comment:
I don't quite understand why you need to extract indices local of face DOFs by `ExtractBdrDofs` (which go in `dofs` here).

If you want to integrate using only their shape functions, you can do it by `FiniteElementSpace::GetFaceElement` and use the result to calculate shape functions only on the face.

Why do you need the local indices of the face DOFs? Have you considered `FiniteElementSpace::GetFaceDofs`?

Reply:
These things are all related to section 2.3 in the draft.
I need to "lump" the flux for each edge/face separately, hence the use of dofs.
GetFaceDofs seems to be working only for continuous elements.

GetFaceElement might be a good option. This part of the code is definitely improvable, but I would like to focus on the Limiter now, so we might have some results for the paper soon.